crypto/dllmain.c

```diff
@@ -25,34 +25,55 @@
  * detaches
  */
 
+static CRITICAL_SECTION csThreadStop;
+static DWORD stop_key;
+static void (*stop_fun)(void *);
+
+int CRYPTO_THREAD_register_stop_function(DWORD key, void (*cleanup)(void *))
+{
+    if (cleanup == NULL)
+        return 1;
+
+    EnterCriticalSection(&csThreadStop);
+    if (stop_fun != NULL) {
+        /* We handle only one stop function. */
+        LeaveCriticalSection(&csThreadStop);
+        return 0;
+    }
+
+    stop_key = key;
+    stop_fun = cleanup;
+    LeaveCriticalSection(&csThreadStop);
+    return 1;
+}
+
+void CRYPTO_THREAD_unregister_stop_function(DWORD key)
+{
+    EnterCriticalSection(&csThreadStop);
+    if (key == stop_key) {
+        stop_key = 0;
+        stop_fun = NULL;
+    }
+    LeaveCriticalSection(&csThreadStop);
+}
+
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved);
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
 {
     switch (fdwReason) {
     case DLL_PROCESS_ATTACH:
-        OPENSSL_cpuid_setup();
```

Comment:
Purpose of this call was to assist legacy API. I mean we still export things like AES_encrypt, RC4, SHAX_Init,Update,Final. Those work without having to perform complete initialization. And without this call they will not live up performance expectations. Now, one can argue that such applications would deserve it. But one can as well argue that since we didn't stop exporting those things, it would be appropriate to maintain it till we do so. At least I'd say that this(*) should be separate commit that would explain implications.

(*) Along with below omitted code.

Reply:
one of the first things OPENSSL_cpuid_setup does call getenv,
and getenv is documented to cause deadlocks when called from DLL_PROCESS_ATTACH.
Google for "DllMain getenv"
or just look into https://rt.openssl.org/Ticket/Display.html?id=2461&user=guest&pass=guest

Reply:
And, in 2009, someone reported a deadlock between _tzset in one thread and
OPENSSL_cpuid_setup call in the DllMain function:
https://social.msdn.microsoft.com/Forums/sqlserver/en-US/b9a18367-41cf-4a1f-8c2a-274bcbef15bb/loader-lock-and-quotenvironmentquot-lock-in-tzset-lead-to-deadlock?forum=windowsgeneraldevelopmentissues

Reply:
As already this *at least* should be separate commit with message that explains why it's done. One can probably wonder if it would be appropriate to solve it other way around, i.e. not remove it, but fix it.

Even OPENSSL_NONPIC_relocated removal should probably be separate commit...

Reply:
Well, since the PR is already approved, I would not want to do that.
However when I removed that code, I only saw a few lines of dead code,
but I was not aware how dangerously wrong it was in fact.
In the light of this discussion, I would prefer to just adjust the commit message to:
```
Fix unexpected call to OPENSSL_thread_stop from DllMain

Instead of directly calling OPENSSL_thread_stop from DllMain
make use of the callback function that is given to the *first*
call to CRYPTO_THREAD_init_local.
At the same time remove the call to OPENSSL_cpuid_setup from DllMain
which can deadlock in getenv.
And remove unused variable OPENSSL_NONPIC_relocated.
```